File: app/api/posts.py
from flask import Blueprint, request, jsonify
from app.models import Toast, Comment, User, db
from flask_login import current_user, login_required
from ..forms.other_forms import ToastForm, CommentForm

posts_routes = Blueprint('posts', __name__)


# Get all posts
@posts_routes.route("")
def get_posts():
    posts = Toast.query.all()
    return {"Posts": [post.to_dict() for post in posts]}


# Get a post by its id
@posts_routes.route("/<int:id>")
def get_post_by_id(id):
    post = Toast.query.get(id)

    if(post):
        return {"Post": post.to_dict()}

    return {"errors": { "message": "Toast Not Found!" } }, 404


# Get all posts by the current user
# @posts_routes.route("/current")
# def get_post_by_user_id():
#     user_id = current_user.id
#     posts = Toast.query.filter(Toast.user_id==user_id).all()

#     return {"Posts": [post.to_dict() for post in posts]}

# Get all posts by a specific user
@posts_routes.route("/specific/<int:id>")
def get_post_by_specific_user_id(id):
    user = User.query.get(id)
    if(user):
        posts = Toast.query.filter(Toast.user_id==id).all()
        return {"Posts": [post.to_dict() for post in posts]}

    return {"errors": { "message": "User Not Found!" } }, 404


# Create a post
@posts_routes.route("", methods=["POST"])
def create_post():
    user_id = current_user.id
    subbreadit_id = request.get_json()["subbreaditId"]

    form = ToastForm()
    form['csrf_token'].data = request.cookies['csrf_token']

    if form.validate_on_submit():
        params = {
            "title": form.title.data,
            "body": form.body.data,
            "user_id": user_id,
            "subbreadit_id": subbreadit_id
        }

        new_toast = Toast(**params)
        db.session.add(new_toast)
        db.session.commit()

        return new_toast.to_dict(), 201

    return form.errors, 400


# Update a post
@posts_routes.route("/<int:id>", methods=["PUT"])
def update_post(id):
    post = Toast.query.get(id)

    form = ToastForm()
    form['csrf_token'].data = request.cookies['csrf_token']

    if (post.user_id == current_user.id):
        if form.validate_on_submit():
            post.title = form.title.data
            post.body = form.body.data

            db.session.commit()

            return post.to_dict(), 201

        return form.errors, 400

    return { "message": "User unauthorized"}, 401



# Delete a post
@posts_routes.route("/<int:id>", methods=["DELETE"])
def delete_post(id):
    post = Toast.query.get(id)
    moderator_id = post.subbreadit.moderator_id

    if (post.user_id == current_user.id or moderator_id == current_user.id):
        db.session.delete(post)
        db.session.commit()

        return {"message": "Success"}, 200

    return { "message": "User unauthorized"}, 401


# Get comment(s) by toast id
@posts_routes.route('/<int:id>/comments')
def get_comments(id):
    comments = Comment.query.filter(Comment.toast_id==id).all()

    return {"Comments": [comment.to_dict() for comment in comments]}


# Create a comment by post id
@posts_routes.route('/<int:id>/comments', methods=["POST"])
def create_comment(id):
    user_id = current_user.id

    form = CommentForm()
    form['csrf_token'].data = request.cookies['csrf_token']

    if form.validate_on_submit():
        params = {
            "body": form.body.data,
            "user_id": user_id,
            "toast_id": id
        }

        new_comment = Comment(**params)

        db.session.add(new_comment)
        db.session.commit()

        return new_comment.to_dict()

    return form.errors, 400

# Create an upvote
@posts_routes.route("/<int:id>/upvote", methods=["POST"])
def create_upvote(id):
    user = User.query.get(current_user.id)
    toast = Toast.query.get(id)

    if toast in user.downvoted_toasts:
        user.downvoted_toasts.remove(toast)
        db.session.commit()

    user.upvoted_toasts.append(toast)
    db.session.commit()

    return toast.to_dict(), 201


# Delete an upvote
@posts_routes.route("/<int:id>/upvote", methods=["DELETE"])
def delete_upvote(id):
    user = User.query.get(current_user.id)
    toast = Toast.query.get(id)

    user.upvoted_toasts.remove(toast)
    db.session.commit()

    return {"message": "Successfully deleted"}, 201

# Create an downvote
@posts_routes.route("/<int:id>/downvote", methods=["POST"])
def create_downvote(id):
    user = User.query.get(current_user.id)
    toast = Toast.query.get(id)

    if toast in user.upvoted_toasts:
        user.upvoted_toasts.remove(toast)
        db.session.commit()

    user.downvoted_toasts.append(toast)
    db.session.commit()

    return toast.to_dict(), 201


# Delete an downvote
@posts_routes.route("/<int:id>/downvote", methods=["DELETE"])
def delete_downvote(id):
    user = User.query.get(current_user.id)
    toast = Toast.query.get(id)

    user.downvoted_toasts.remove(toast)
    db.session.commit()

    return {"message": "Successfully deleted"}, 201

File: react-vite/src/redux/posts.js
import { CREATE_DOWNVOTE, CREATE_UPVOTE, DELETE_DOWNVOTE, DELETE_UPVOTE } from "./votes";

export const LOAD_POSTS = 'posts/LOAD_POSTS';
export const LOAD_POST = 'posts/LOAD_POST';
export const CREATE_POST = 'posts/CREATE_POST';
export const UPDATE_POST = 'posts/UPDATE_POST';
export const REMOVE_POST = 'posts/REMOVE_POST';

export const loadPosts = (posts) => ({
  type: LOAD_POSTS,
  posts
});

export const loadPost = (post) => ({
  type: LOAD_POST,
  post
});

export const createPost = (post) => ({
  type: CREATE_POST,
  post
});

export const editPost = (post) => ({
  type: UPDATE_POST,
  post
});

export const removePost = (postId) => ({
  type: REMOVE_POST,
  postId
});

export const getPosts = () => async dispatch => {
  const response = await fetch(`/api/posts`)

  if (response.ok) {
    const posts = await response.json()
    dispatch(loadPosts(posts))
  } else {
    const errors = await response.json()
    return errors
  }
}

export const getPostById = (postId) => async dispatch => {
  const response = await fetch(`/api/posts/${postId}`)

  if (response.ok) {
    const post = await response.json()
    dispatch(loadPost(post))
    return post
  } else {
    const errors = await response.json()
    return errors
  }
}

export const getPostsByUserId = () => async dispatch => {
  const response = await fetch(`/api/posts/current`)

  if (response.ok) {
    const post = await response.json()
    dispatch(loadPosts(post))
    return response
  } else {
    const errors = await response.json()
    return errors
  }
}

export const getPostsBySpecificUserId = (userId) => async dispatch => {
  const response = await fetch(`/api/posts/specific/${userId}`)

  if (response.ok) {
    const posts = await response.json()
    dispatch(loadPosts(posts))
    return response
  } else {
    const errors = await response.json()
    return errors
  }
}

export const getSubbreaditPosts = (subbreaditId) => async dispatch => {
  const response = await fetch(`/api/subbreadits/${subbreaditId}/posts`)

  if (response.ok) {
    const posts = await response.json()
    dispatch(loadPosts(posts))
  } else {
    const errors = await response.json()
    return errors
  }
}

export const addPost = (post) => async dispatch => {
  const response = await fetch(`/api/posts`, {
    method: "POST",
    body: post
  })

  if (response.ok) {
    const post = await response.json()
    dispatch(createPost(post))
    return post
  } else {
    const errors = await response.json()
    return {"errors": errors}
  }
}

export const updatePost = (postId, post) => async dispatch => {
  const response = await fetch(`/api/posts/${postId}`, {
    method: "PUT",
    body: post
  })

  if (response.ok) {
    const post = await response.json()
    dispatch(editPost(post))
    return post;
  } else {
    const errors = await response.json()
    return {"errors": errors}
  }
}

export const deletePost = (postId) => async dispatch => {
  const response = await fetch(`/api/posts/${postId}`, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json"
    },
  })

  if (response.ok) {
    dispatch(removePost(postId))
  } else {
    const errors = await response.json()
    return errors
  }
}

const postsReducer = (state = {}, action) => {
  switch (action.type) {
    case LOAD_POSTS: {
      const postsState = {};
      if (action.posts.Posts.length) {
        action.posts.Posts.forEach((post) => {
          postsState[post.id] = post;
        });
      }
      return postsState;
    }
    case LOAD_POST: {
      const postsState = {};
      const post = action.post.Post
      postsState[post.id] = post
      return postsState;
    }
    case CREATE_POST:
      return { ...state, [action.post.id]: action.post };
    case UPDATE_POST:
      return { ...state, [action.post.id]: action.post };
    case REMOVE_POST: {
      const newState = { ...state };
      delete newState[action.postId];
      return newState;
    }
    case CREATE_UPVOTE:
    return { ...state, [action.toast.id]: { ...state[action.toast.id], upvotes: [ ...state[action.toast.id].upvotes, action.userId], downvotes: state[action.toast.id].downvotes.filter(id => id != action.userId) }}
    case CREATE_DOWNVOTE:
      return { ...state, [action.toast.id]: { ...state[action.toast.id], downvotes: [ ...state[action.toast.id].downvotes, action.userId], upvotes: state[action.toast.id].upvotes.filter(id => id != action.userId) }}
    case DELETE_UPVOTE:
      return { ...state, [action.toastId]: { ...state[action.toastId], upvotes: state[action.toastId].upvotes.filter(id => id != action.userId)}}
    case DELETE_DOWNVOTE:
      return { ...state, [action.toastId]: { ...state[action.toastId], downvotes: state[action.toastId].downvotes.filter(id => id != action.userId)}}
    default:
      return state;
  }
};

export default postsReducer;
